### DISCUSSION TITLE
Import module without importing there SCons manually?

### DISCUSSION THREAD

**Discussion by ghost** [SCons/scons#4654] [Q&A]
Hello! When you run `scons` your `SConstruct` file doesn't need to explicitly load all the `SCons` modules. The same is with all other files called through `SConscript` but if I import another module I need to explicitly import `SCons` parts which I need.

 Is there any function like `Import("module")`? 
(I know `Import` name is already occupied just an example). I know there is `Export` and `Import` which allows to pass values between files called with `SConscript` but I need to import everything from there. I want to replace a `import module` to `Import("module")` and do then `module.my_func()` like it was just normal module loading. Is there something like it? Or do you have any ideas how to do it in another way?

**Comment by bdbaddog** [comment]
If you look in the manpage, you'll see the following:

> Builder methods that can be called without an explicit environment (indicated in the listing of builders below without a leading env.) may be called from custom Python modules that you import into an SConscript file by adding the following to the Python module:
> 
> ```
> from SCons.Script import *
> ```

It's also mentioned in the user guide here: https://scons.org/doc/production/HTML/scons-user.html#id1462

**Reply by ghost** [reply]
I will assume it as "no"

**Reply by bdbaddog** [reply]
Was the above not an answer to your question?

Seems maybe I misunderstood?
What are you trying to do then?
Not use python's `import` to import a python module?

**Reply by ghost** [reply]
Just formally yeah. Putting `from SCons.Script import *` allows to use SCons in a module (which I need) but I've asked is there a way to load a module with builtin SCons function something like `SConsModule("module_name")` which will load the module and automatically "put" that `from SCons.Script import *` like it is done with `SConscript`

**Reply by bdbaddog** [reply]
If you loaded it via `SConscript()`, though. then you'd need to Export()/Import() to get info from the "parent" sconscript.

**Reply by ghost** [reply]
I meant `SConscript` which will do something like `Return()` everything defined in module + automatically putting there `from SCons.Script import *`

**Reply by bdbaddog** [reply]
Not sure I understand, if that's what you want, why not just use SConscript() to load the logic?

**Reply by ghost** [reply]
OK. How to return everything defined in module? I have a `Return()` which needs names...

**Reply by bdbaddog** [reply]
What do you want to define inside this file and return? 
are you calling builders? Configuring Environment()?

**Reply by ghost** [reply]
Lots of functions some of which are SCons related, some of which are not... Some can be considered configuring environment, some can be seen as builders, some can be seen as not very related functions

**Reply by bdbaddog** [reply]
I'd just put the logic in `site_scons` have it use `from SCons.Script import *`, that's likely much cleaner and far more standard usage.

Go read this bit and see if that's not really better..
https://scons.org/doc/production/HTML/scons-user.html#id1462

**Reply by bdbaddog** [reply]
I'm taking about site_scons/site_init.py (and anything it imports)...

**Reply by ghost** [reply]
As I sad `from SCons.Script import *` works and `site_scons` is a cool thing which I plan to adopt.... Or maybe I didn't mention explicitly `from SCons.Script import *` works for me if so sorry :sweat_smile: 

---

The current situation is following. There is a module. Which is loaded dynamically by name (determined by user input) and I need to load only this specific one.

```python
#main.py
sys.path.insert(<path_to_module>)
import module
module.method1()
module.method2()
....
```

---

But there is `SConscript()` which allows to load any file and has a `Return()`. But I need to explicitly `Return()` everything defined in module... Think of it like almost normal python module (which occasionally uses SCons features) In this case I am getting boilerplate inside every module but clean main file

```python
#main.py

module = SConscript("path_to_module")
module.method1()
module.method2()
...
```

```python
# in each module
from SCons.Script import *

# modules content

Return(<everything>) # I even don't know how to do it...
```

---

What I am talking about is new method

```python
SConsmodule("path_to_module", "loaded_module_name")

loaded_module_name.method1()
loaded_module_name.method2()
```

This `SConsmodule` function does the same as `SConscript` but explicitly returns everything defined in module like normal `import module` would do. In general it is something in between `SConscript` and `import`

**Reply by bdbaddog** [reply]
No such thing. And we honeslty won't implement. 
That's what site_scons/site_init.py is for.
Anything there is in scope of the SConscripts..
I think that's likely what you're shooting for?

Otherwise if not referenced in site_scons/site_init.py, `import xyz` will look in `site_scons`..

**Reply by ghost** [reply]
Many people (which I asked) has opinion about SCons: "It doesn't force you to do almost anything. You can anytime do a workaround and do thing by yourself... Not using builders and etc... But if you want to migrate from not very SConsish codebash to the one which uses SCons in more correct way you can need something like it... I think it is unnecessary for a fully SConsish codebase but for the one which does a migrations it can be a handful thing

**Reply by ghost** [reply]
But do you at least know how to return everything?

**Reply by bdbaddog** [reply]
just set a package variable and then you can reference it.

```
import mystuff # This would be site_scons/mystuf.py

x=mystuff.some_stored_value
```

# mystuff.py
```
some_stored_value="Happy Wednesday"
```


Yes. You can do whatver you want with SCons. Somethings will not be wise, will be fragile, and hard to track down problems with.

If you go the most mainstream usage (as with any tool), there's likely to be more information,testing, and help available.

**Reply by mwichmann** [reply]
"Everything" is a big word, but is you want to use the Import/Export mechanism, put everything you want to share in a container, and just export that one name. In fact, the fairly obvious container is... a construction environment.

**Comment by mwichmann** [comment]
Tool modules have funny importing under the covers, if their particular approach happens to fit your use case, you might be able to make use of a tool.
